Use and combine chain data with LSQ data for stake pools

Anviking opened the pull request:
# Issue Number

ADP-311, #1719 

# Overview

- [x] Add Shelley `MsgFetchedNodePoolLsqData` trace
- [x] Use and combine chain data with LSQ when listing pools
- [x] Update swagger to make pool cost and margin nullable
- [x] Make API types nullable
- [x] Add model implementation for `readPoolMetadata`


# Comments

- [ ] We should test that we order by non-myopic rewards, but that would require the pools producing blocks. 
- [ ] Maybe we should have a DB test checking that `readMetadata` returns what we previously put in the DB.

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
Add Shelley `MsgFetchedNodePoolLsqData` trace

Anviking committed:
Read and combine DB data with LSQ when listing pools

We already have the workers to pool registration certificates and
metadata in the data in the DB. Here we read it.

- Ignore blocks produced by BFT nodes

- Relocate functions and add documentation

- Use strategy `chainButNoLsq = dropMissing`

- Check that pools produce blocks

- fixup: combineDBAndLsqData -> combineDbAndLsqData

- fixup: rewrite fetching of certMap

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -417,58 +429,80 @@ spec = do
             ]
 
     it "STAKE_POOLS_LIST_01 - List stake pools" $ \ctx -> do
-        eventually "Listing stake pools shows expected information" $ do
```
I think there's a _theoretical_ chance that the pool registration certificates haven't been discovered yet, but I'm not sure it's possible in practice.

It fails quicker without the default delay.

I guess we could use `eventuallyUsingDelay (5 * second)` also.

Anviking commented on lib/core/src/Cardano/Pool/DB.hs:
```diff
@@ -113,10 +113,13 @@ data DBLayer m = forall stm. (MonadFail stm, MonadIO stm) => DBLayer
     , unfetchedPoolMetadataRefs
-        :: stm [(PoolId, StakePoolMetadataUrl, StakePoolMetadataHash)]
+        :: Int
+        -> stm [(PoolId, StakePoolMetadataUrl, StakePoolMetadataHash)]
```
🧐 bad rebase

Anviking committed:
Allow pool cost and margin to be ommitted in swagger.yaml

If their pool registration hasn't been found yet the fields will be left out from the response.

Anviking committed:
Add readPoolMetadata DB function

Anviking committed:
Add MsgWillOpenDB log message

Anviking committed:
Log static metadata server urls when launched

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -416,68 +428,108 @@ spec = do
+                        , expectListField 1
+                            (#metrics . #producedBlocks) (.>= Quantity 0)
+                        , expectListField 2
+                            (#metrics . #producedBlocks) (.>= Quantity 0)
```
Now after rebasing I see them producing blocks, but I still see no rewards 🤔 

_or I wasn't fully rebased… trying again_

_I think some monetary protocol parameters need to be tweaked in the genesis file_

KtorZ replied:
> I think some monetary protocol parameters need to be tweaked in the genesis file

That is possible.

Anviking replied:
I created the related https://github.com/input-output-hk/cardano-wallet/pull/1783, which might be good for investigating. We can leave the todo for rewards in this PR for now.

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -416,68 +428,108 @@ spec = do
+    let fixTypeInference = True `shouldBe` True
+    let listPools ctx = request @[ApiStakePool] @IO ctx (Link.listStakePools arbitraryStake) Default Empty
+    describe "STAKE_POOLS_LIST_01 - List stake pools" $ do
+            it "immediately has non-zero saturation & stake" $ \ctx -> do
```
neat-picking but why the double indentation here?

Anviking replied:
A remnant, fixed.

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -416,68 +428,108 @@ spec = do
+                        , expectField id $ \pools ->
+                            -- To ignore the arbitrary order,
+                            -- we sort on the names before comparing
+                            sortOn name ( mapMaybe (fmap getApiT . view #metadata) pools)
```
:+1:

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -416,68 +428,108 @@ spec = do
+                                , description = Just "Lorem Ipsum Dolor Sit Amet."
+                                , homepage = "https://iohk.io"
+                                }
+                            ]
```
Maybe we could avoid hard-coding the list here since it's available as `operators` in `Cardano.Wallet.Shelley.Launch`.

Anviking replied:
Don't think we can (or want) `cardano-wallet-core-integration` depend on `cardano-wallet-shelley`… :S  We _could_ move `API.Shelley.StakePools` to `cardano-wallet-shelley:integration` though?

KtorZ replied:
> cardano-wallet-core-integration depend on cardano-wallet-shelley

Arf. I didn't realize that. That's quite lame indeed.

KtorZ commented on lib/core/src/Cardano/DB/Sqlite.hs:
```diff
@@ -463,6 +465,7 @@ instance ToText DBLog where
         MsgQuery stmt _ -> stmt
         MsgRun False -> "Running database action - Start"
         MsgRun True -> "Running database action - Finish"
+        MsgWillOpenDB fp -> "Will open db at " <> (maybe "in-memory" T.pack fp)
```
:+1: Anything particular you were looking for :thinking: ?

Anviking replied:
I wanted to browse manually to check 1) if pool registration certificates had metadata references 2) if any metadata had been fetched

I believe I was simply missing an `eventually` though.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs:
```diff
@@ -440,6 +440,7 @@ withStakePool tr idx cfg action =
     bracketTracer' tr "withStakePool" $
         withTempDir tr name $ \dir -> do
             withStaticServer dir $ \url -> do
+                putStrLn $ "Launching static server at url: " <> url
```
:grimacing: ... There's litteraly a tracer `tr` in scope that can be used instead of printing to stdout!

Anviking replied:
Fixed d31e78541651f28bc38b04be24f3122c1712e226

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -98,13 +111,130 @@ import qualified Data.Map.Merge.Strict as Map
+        tip <- getTip
+        let dummyCoin = Coin 0
+        res <- runExceptT $ map fst . Map.toList
+            . combineLsqData <$> stakeDistribution nl tip dummyCoin
```
known pools should be implemented using the _database_ only. That is its sole purpose, otherwise it is utterly redundant with listPools. The goal is to have a fast access to known pools ids to avoid having to query the network layer when delegating / undelegating to pools.

Anviking replied:
What do you say about: https://github.com/input-output-hk/cardano-wallet/pull/1753#discussion_r440780076 ?

> I know knownPools are used to validate the users's choice of delegation.
>
> The pool worker is less likely to be in sync with the network (more downstream) than the node.
>
> Wouldn't implementing it in terms of the LSQ make more sense?

Though perhaps, having the pool worker out of sync with the node is unlikely, since they Daedalus users will always launch the node and wallet at the same time?

KtorZ replied:
Because that is why we have this function in the first place. We can already get the full list of pool ids from the `listPools` and it was how it was done _before_. Yet, this was slow and sometimes _unreliable_ because it needed to go through much more work. As a result, requests for joining / quitting pools were often failing because we couldn't reliably get a list of pools ids. 

Yet, getting the list of pool ids is straightforward and very reliable when fetched from the local database. Implementing this function using the LSQ completely defies its original purpose which was to circumvent the network layer!

Anviking replied:
In jormungandr we required the DB to be in sync with the node. For haskell we don't.

If the user ends up with a synced node and no stake-pools.sqlite, and lists pools, he will see pools based on LSQ data from the node tip. If he were to try to delegate, I fear the wallet wouldn't let him, claiming that the pool he selected doesn't exist.

It just seems weird to me that we would treat the node as the source of [truth](https://github.com/input-output-hk/cardano-wallet/blob/2f1a21f67f1ab3f5ae8e57da5c97842bf0d0532d/lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs#L201), but then validate the delegation choice based a different one.

KtorZ replied:
> If he were to try to delegate, I fear the wallet wouldn't let him, claiming that the pool he selected doesn't exist.

Which is fine, the pool worker won't be _that_ behind and will catch up quite quickly. We can leave it as it is; if the networking layer is fast enough and reliable, then it _should be okay_. Let's see.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -98,13 +111,130 @@ import qualified Data.Map.Merge.Strict as Map
+              . sortOn (Down . (view (#metrics . #nonMyopicMemberRewards)))
+              . map snd
+              . Map.toList
+              $ combineDBAndLsqData lsqData chainData
```
_You can ignore this comment_

The mixed case style somewhat triggers some OCDs in me. 

```
combineDBAndLSQData
```

or 

```
combineDbAndLsqData
```

Mixing both looks weird.

Anviking replied:
Went with `combineDbAndLsqData`. I can't find a guideline [here](https://github.com/input-output-hk/adrestia/wiki/Coding-Standards), but I thought `Lsq` was preferred to `LSQ`.

We do use `DB` a lot in our code.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -98,13 +111,130 @@ import qualified Data.Map.Merge.Strict as Map
+    -- TODO: Maybe we have simply missed a retirement certificate. Maybe we
+    -- should simply dropMissing here.
+    chainButNoLsq = traverseMissing $ \k _ ->
+        error $ "Pool exists in DB but not in LSQ: " <> show k
```
Yes we should! We do not track retirement certificates yet, so we definitely need this to account for them, same as we did for the ITN. When there's a pool in our database, but not returned by the LSQ, we fully trust the node and discard the pool on our end :+1:

Anviking replied:
Ok, will change now

Anviking replied:
Fixed.

rvl commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -27,11 +27,23 @@ import Cardano.Wallet.Primitive.AddressDerivation
 import Data.Quantity
     ( Quantity (..) )
+import Data.Sort
+    ( sortOn )
```
You don't need the sort package. It's in base `Data.List.sortOn`.

rvl commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs:
```diff
@@ -1038,6 +1039,7 @@ withSystemTempDir tr name action = do
 data ClusterLog
     = MsgCartouche
     | MsgLauncher String LauncherLog
+    | MsgServerForMetadataStarted FilePath
```
Might be good to log both the directory and url - the url type being String.

Anviking replied:
👍 now
```
cardano-wallet.integration:Info:43] [2020-06-23 09:01:53.84 UTC] Started a static server for http://localhost:60431/ at /private/var/folders/h_/mlzym27d5lz_fx3ph2d547700000gn/T/integration-62aa8274a6e3655b/stake-pool-2
[cardano-wallet.integration:Info:41] [2020-06-23 09:01:53.84 UTC] Started a static server for http://localhost:60430/ at /private/var/folders/h_/mlzym27d5lz_fx3ph2d547700000gn/T/integration-62aa8274a6e3655b/stake-pool-0
[cardano-wallet.integration:Info:46] [2020-06-23 09:01:53.84 UTC] Started a static server for http://localhost:60432/ at /private/var/folders/h_/mlzym27d5lz_fx3ph2d547700000gn/T/integration-62aa8274a6e3655b/stake-pool-1
```

rvl commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -123,76 +253,53 @@ combineLsqData NodePoolLsqData{nOpt, rewards, stake} =
+readDBPoolData DBLayer{..} = atomically $ do
+    pools <- listRegisteredPools
+    certMap <- Map.fromList . catMaybes <$> forM pools
+        (\poolId -> ((fmap (poolId,)) <$> readPoolRegistration poolId))
```
A bit simpler...

```suggestion
    registrations <- mapM readPoolRegistration pools
    let certMap = Map.fromList [(poolId, cert) | (poolId, Just cert) <- zip pools registrations]
```

Anviking replied:
`zip` feels a bit wrong to me, but sure, it does look neat

rvl reviewed (commented):
Looks decent

KtorZ commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -415,68 +426,107 @@ spec = do
-        eventually "Listing stake pools shows expected information" $ do
-            r <- request @[ApiStakePool] ctx
-                (Link.listStakePools arbitraryStake) Default Empty
+    let fixTypeInference = True `shouldBe` True
```
You just need a `@IO` down below on `request`

KtorZ commented on lib/shelley/test/data/cardano-node-shelley/genesis.yaml:
```diff
@@ -6,7 +6,7 @@ protocolParams:
     minor: 0
     major: 0
-  decentralisationParam: 0.97 # means 3% decentralised
+  decentralisationParam: 0.1 # means 90% decentralised
```
This was very high to avoid too frequent rollbacks until we have a proper transaction scheduler. I don't think you'll get anything merged with 90% decentralization :)

KtorZ replied:
We can try, but I am afraid this is going to block many PRs after that :/

KtorZ committed:
use type applications instead of 'fixTypeInference' function

KtorZ committed:
revert changing decentralizationParam to 0.1

We need to keep this one quite high at the moment in order to prevent the cluster from rolling back too often. The wallet can handle rollback quite nicely, but the integration scenarios cannot (most of them). A lot of scenarios are built following a basic fixture -> action -> assertions principle, and rolling back the fixture can have very negative effects on the rest. We could run every scenario multiple times (at least twice) to maybe cope with intermediate failures due to rollback...

KtorZ committed:
split assertions on block production and saturation out in a separate scenario

So that we don't mix concerns between scenarios and we can lower down a bit the assertions without loosing much of the testing benefits.

KtorZ reviewed (approved):
reviewed, tweaked a few things :+1:
